### PR TITLE
Com 33

### PR DESCRIPTION
**Pull request by teckays**

### CODE REVIEW

File: lib/sender.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    })
+
+    if (options.batchSize < 1 || options.batchSize > 10) {
+      throw new Error(`SQS batch size must be between 1 and 10. [${options.batchSize}] specified.`)
+    }
+  }
+
+  /**
+   * Generate
+   * @returns {*}
```
Comment (pavelpichrt) [inline lib/sender.js]: returns string

File: lib/sender.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+   */
+  static messageId () {
+    return UUID.v4()
+  }
+
+  /**
+   * Create message attribute
+   * @param {string} type - Attribute data type
+   * @param {String|Buffer|Array<String>|Array<Buffer>} value
+   * @returns {{DataType: *}}
```
Comment (pavelpichrt) [inline lib/sender.js]: is this a valid jsdoc? (not sure myself)

File: lib/sender.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    }
+  }
+
+  /**
+   * Validate Object message attribute
+   * @param {Object} message - Message object
+   * @returns {{Id: (*|String|string), MessageBody: HTMLElement}|*}
+   * @private
+   */
+  _validateObjectMessage (message) {
```
Comment (pavelpichrt) [inline lib/sender.js]: I would expect "validateXXX" to return boolean, this looks more like "formatXXX"

File: lib/sender.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    return entry
+  }
+
+  /**
+   * Prepare message for SQS
+   * @param {string} message - Message body
+   * @returns {{Id: *, MessageBody: string}}
+   * @private
+   */
+  _prepareMessage (message) {
```
Comment (pavelpichrt) [inline lib/sender.js]: What does it mean "prepare"? Looks like another "format" kind of function?

File: lib/sender.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    throw new Errors.InvalidMessage(`Message type must be string, [${typeof message}] given`)
+  }
+
+  /**
+   * Prepare SQS entries
+   * @param {Array} messages - A message or an array of messages
+   * @returns {Promise.<Array>}
+   * @private
+   */
+  _prepareEntries (messages) {
```
Comment (pavelpichrt) [inline lib/sender.js]: What is SQS entry? :) Tried to google it without luck

File: lib/sender.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    return new Promise((resolve, reject) => {
```
Comment (pavelpichrt) [inline lib/sender.js]: Why use a promise here?

File: lib/sender.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+      })
+    })
+  }
+
+  /**
+   * Create a String message attribute object
+   * @param {String} value - Attribute value
+   * @returns {{DataType}|{DataType: *}}
+   */
+  stringAttribute (value) {
```
Comment (pavelpichrt) [inline lib/sender.js]: I think this should be something like addStringAttribute?

File: lib/sender.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+    if (messages.length > this.batchSize) {
+      Promise.reject(new Errors.InvalidMessage(`Message batch must not be greater than ${this.batchSize}.`))
+    }
+
+    Debug(`Sending [${messages.length}] messages ...`)
+
+    return this
+      ._prepareEntries(messages)
+      .then(entries => this._sendBatch(entries))
```
Comment (pavelpichrt) [inline lib/sender.js]: .then(this._sendBatch)

File: tmp/test.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+'use strict'
```
Comment (pavelpichrt) [inline tmp/test.js]: forgotten file? :)

File: lib/sender.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+    // Instantiate SQS
+    this.sqs = options.sqs || new AWS.SQS({
+      region: options.region || 'eu-west-1'
+    })
+  }
+
+  /**
+   * Validate Sender required options
+   * @param options
```
Comment (pavelpichrt) [inline lib/sender.js]: missing param type

File: lib/sender.js (hunk 5 above)
Comment (teckays) [inline lib/sender.js]: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SQS.html#sendMessageBatch-property

it's basically a property of the whole object passed to `AWS.SQS`, see the link above:
```
var params = {
  Entries: [...], // this is the one
  ...
}
```

File: lib/sender.js (hunk 6 above)
Comment (teckays) [inline lib/sender.js]: I'm throwing an error inside the `_formatMessage()` method and using a promise to be able to send the error back to the `Sender.send().catch(...)` call.

File: lib/sender.js (hunk 8 above)
Comment (teckays) [inline lib/sender.js]: context problem, (an old one we discussed). `this` inside `_sendBatch()` is not the instance of my class but a `Promise` instance instead.

Comment (pavelpichrt) [inline lib/sender.js]: ok, makes sense

**Pull request merged**